[#8469] IterableOnce#copyToArray returns number of elements copied

joshlemer opened the pull request:
This fixes scala/bug#8469, and also adds `StringOps#copyToArray(Array[Char])`

This also fixes scala/bug#11048

NthPortal commented:
@joshlemer I'm marking this as WIP because you said it's not ready for merging. Just drop a message here when that changes, and someone can remove the label :)

julienrf commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -729,24 +730,27 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
+      written += 1
     }
-    xs
+    written
```
We could also write `i - start`, and remove the `written` variable. I’m not sure which solution is better, though.

joshlemer replied:
👍

julienrf reviewed (commented):
Thanks @joshlemer, I think this is the right way to go!

The fact that you can not call `copyToArray` on a `String` is a regression. I’ve just noticed that one overload of `copyToArray` is missing. You can copy the way it is defined in `IterableOnceOps`.

julienrf commented on src/library/scala/collection/ArrayOps.scala:
```diff
@@ -1263,13 +1263,18 @@ final class ArrayOps[A](val xs: Array[A]) extends AnyVal {
-    Array.copy(xs, 0, dest, start, min(xs.length, min(dest.length-start, len)))
-    dest
+  def copyToArray[B >: A](dest: Array[B], start: Int, len: Int = Int.MaxValue): Int = {
+    val copied = min(xs.length, min(dest.length-start, len))
```
The maths are not correct in case the `start` parameter is higher than `dest.length`. Consider the case where `start = dest.length + 1` (and `len` is a positive number):

~~~
copied = min(xs.length, min(dest.length - start, len))
copied = min(xs.length, min(-1, len)
copied = min(xs.length, -1)
copied = -1
~~~

julienrf commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -1198,7 +1201,11 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
   def toArray[B >: A: ClassTag]: Array[B] =
-    if (knownSize >= 0) copyToArray(new Array[B](knownSize), 0)
+    if (knownSize >= 0) {
+      val a = new Array[B](knownSize)
```
Do you mind giving `a` a more meaningful name? e.g. `copied` or `written`.

joshlemer replied:
👍

julienrf commented on src/library/scala/collection/StringOps.scala:
```diff
@@ -1142,11 +1140,25 @@ final class StringOps(private val s: String) extends AnyVal {
+    */
+  def copyToArray(xs: Array[Char], start: Int, len: Int): Int = {
+    val written = max(min(min(s.length, len), xs.length - start), 0)
+    s.getChars(0, written, xs, start)
```
We could event put this line within an `if (written > 0)`

joshlemer replied:
👍

julienrf commented on src/library/scala/collection/immutable/ArraySeq.scala:
```diff
@@ -133,12 +133,14 @@ sealed abstract class ArraySeq[+A]
-    if(l > 0) Array.copy(unsafeArray, 0, xs, start, l)
-    xs
+  override def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): Int = {
+    val _len = math.max(math.min(scala.math.min(len, length), xs.length-start), 0)
```
I suggest that we use the same identifier everywhere. E.g. `copied`?

joshlemer replied:
👍

julienrf commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -159,12 +159,14 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
-    if(l > 0) Array.copy(array, 0, xs, start, l)
-    xs
+  override def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): Int = {
+    val _len = scala.math.min(scala.math.min(len, length), xs.length-start)
```
Also, I’m thinking that it could be beneficial to define only once the computation that returns the number of copied elements, and then reuse it everywhere so that we don’t miss the final `max(…, 0)`. You could, for instance, create a private method in the `IterableOnce` companion object.

julienrf replied:
We could even put the `if (copied > 0)` in that reusable method to save some logic duplication…

~~~ scala
@inline private[collection] def copyToArrayTemplate[A](
    thisLength: Int, xs: Array[A], start: Int, len: Int)(doCopy: Int => Unit): Int = {
  val copied = math.max(math.min(scala.math.min(len, thisLength), xs.length-start), 0)
  if (copied > 0) {
    doCopy(copied)
  }
  copied
}
~~~

And then, at use site:

~~~ scala
override def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): Int =
  copyToArrayTemplate(length, xs, start, len) { copied =>
    Array.copy(array, 0, xs, start, copied)
  }
~~~

WDYT?

(You can use a better name than `copyToArrayTemplate`, obviously…)

(And I’m not sure how efficiently are literal functions inlined/optimized in such cases)

joshlemer replied:
Hey @julienrf I really like the idea of implementing this a single time. I think though that the utility method could just return the number of elements to write, and that would save us having to allocate a closure

julienrf reviewed (commented):
Thanks @joshlemer! The implementations or overrides are not always consistent together. I’ve left a comment with a suggestion to factor them out, let me know what you think!

NthPortal commented:
@joshlemer do you wanna edit the PR description to no longer say that it's not ready for review, btw?

joshlemer commented:
@NthPortal done, not sure why the build fails, it compiles fine for me

julienrf reviewed (approved):
Thanks! Can you squash your commits?

joshlemer commented:
@julienrf yep! I'll do this tonight probably

joshlemer committed:
[#8469] IterableOnce#copyToArray returns number of elements copied (Int)

[#11048] Array and ArrayDeque copyToArray do not throw IndexOutOfBoundException

[11048] Add junit tests for {Queue,mutable.Queue}.copyToArray

joshlemer commented:
@julienrf commits are squashed

dwijnand commented on test/junit/scala/collection/IterableTest.scala:
```diff
@@ -55,124 +56,116 @@ class IterableTest {
-    assert(Map.empty      == mutable.OpenHashMap.empty.toTraversable)
-    assert(Set.empty      == mutable.Set.empty.toTraversable)
+    assert(Map.empty == mutable.OpenHashMap.empty.toTraversable)
+    assert(Set.empty == mutable.Set.empty.toTraversable)
```
This is a little sad.. 😕

NthPortal replied:
+1, I would like to keep the alignment

joshlemer replied:
:+1: will fix

joshlemer replied:
@dwijnand @NthPortal btw this is fixed now

szeiger commented on src/library/scala/collection/immutable/ArraySeq.scala:
```diff
@@ -5,6 +5,7 @@ import java.io.{ObjectInputStream, ObjectOutputStream}
 
 import scala.collection.mutable.{ArrayBuffer, ArrayBuilder, Builder, ArraySeq => MutableArraySeq}
 import scala.collection.{ArrayOps, ClassTagSeqFactory, SeqFactory, StrictOptimizedClassTagSeqFactory, View}
+import scala.collection._
```
This appears to be what causes the build errors. Never import `collection._` in a more specific collection package. You want to use the names from `immutable`, in particular the build errors are due to `immutable.ArraySeq` now extending `collection.IndexedSeq` instead of `immutable.IndexedSeq`.

joshlemer commented:
I believe this is ready for merging now :-)
